src/lib.rs

```diff
@@ -0,0 +1,320 @@
+#![crate_type = "staticlib"]
+
+use libc::{c_char, c_int, c_void, size_t, ssize_t};
+use std::io::{Cursor, Read, Write};
+use std::slice;
+use std::{ffi::CStr, sync::Arc};
+use std::{io::ErrorKind::ConnectionAborted, mem};
+
+use rustls::{ClientConfig, ClientSession, Session};
+
+type CrustlsResult = c_int;
+
+pub const CRUSTLS_OK: c_int = 0;
+pub const CRUSTLS_ERROR: c_int = 1;
+
+#[no_mangle]
+pub extern "C" fn rustls_client_config_new() -> *const c_void {
+    let mut config = rustls::ClientConfig::new();
+    config
+        .root_store
+        .add_server_trust_anchors(&webpki_roots::TLS_SERVER_ROOTS);
+    env_logger::init();
+    Arc::into_raw(Arc::new(config)) as *const c_void
```

Comment:
Oh, wait, I think I misunderstood how `into_raw` and `from_raw` work: the `Arc` and its reference count aren't thrown away, they are still in the buffer passed back to C-land. But suppose you have two threads, and one of them calls `rustls_client_config_free` at the same time that the other one calls `rustls_client_session_new` with the same pointer. The two functions will separately use `Arc::from_raw` to recreate an `Arc`. Will those two distinct `Arc`s give you the runtime mutual exclusivity you want? That is, is calling `Arc::from_raw` on the same raw buffer twice the same as doing `std::clone::clone()` on an `Arc`?

Reply:
You're correct that we're not getting any of the semantics of Arc while the pointer is owned by C-land. We don't get the reference counting, let alone the locking. The reason I'm using an Arc rather than an Rc or a Box is that rustls' ClientSession::new requires an Arc as part of its type signature. Also, we need C's copy of the pointer to remain valid long-term, so we can't do `Arc::new(ptr)` before each call to ClientSession::new, because that would move the value pointed to by ptr, invalidating C's copy.

Also a small note: When we do `Arc::into_raw`, we actually get a pointer to the inner value (i.e. `ClientConfig`). The strong and weak counters stay in memory right where they are, a few bytes lower than `ClientConfig`. When we do `from_raw`, it does a negative offset to reconstruct the Arc, including those strong and weak counters that were hiding out at a lower memory location.

Reply:
Ok, so presumably the same goes for the mutex protecting those counters, so two separate `Arc`s that were `from_raw`ed into existence will be using the same lock and incrementing/decrementing the same reference count(s). Wow, that's pretty cool!

Reply:
Yes, though to be pedantic, there's no lock, just atomic::AtomicUsize counts: https://doc.rust-lang.org/src/alloc/sync.rs.html#281-294